service/src/main/java/bio/terra/externalcreds/services/ProviderService.java

```diff
@@ -164,6 +169,47 @@ public void deleteLink(String userId, String providerId) {
     linkedAccountService.deleteLinkedAccount(userId, providerId);
   }
 
+  @VisibleForTesting
+  void authAndRefreshPassport(LinkedAccount linkedAccount) {
+    if (linkedAccount.getExpires().before(Timestamp.from(Instant.now()))) {
+      passportService.deletePassport(linkedAccount.getId().get());
+      linkedAccountService.updateLinkAuthenticationStatus(linkedAccount.getId().get(), false);
+    } else {
+      try {
+        var clientRegistration =
+            providerClientCache.getProviderClient(linkedAccount.getProviderId());
+        var accessTokenResponse =
+            oAuth2Service.authorizeWithRefreshToken(
+                clientRegistration.orElseThrow(),
+                new OAuth2RefreshToken(linkedAccount.getRefreshToken(), null));
+        var userInfo =
```

Comment:
If a new refresh token is returned in the `accessTokenResponse` it should be saved here, before the call to `getUserInfo` because if that call fails we lose the new refresh token and the link becomes unauthenticated because the original refresh token becomes invalid.

Reply:
Good point, thanks! Just updated that.